### DISCUSSION TITLE
Data wont load in preview

### DISCUSSION THREAD

**Discussion by d-opri** [neuefische/questions#505] [Web] [answered]
moin moin,

i managed to get my bills displayed on individual details pages, however there are a few things not working: 1. the balance of the participants wont show, intitally the participant object in my form had the values name and paid, after calculation is triggered there should be a new balance value in the object. it also doesnt show in console log. 2. some of the bills open to the type error "bill is null". 3. now i tried to use it in the vercel preview and the bill wont even load. 

in my form component, the calculation:
```
function handleSubmit(event) {
    event.preventDefault();
    const fairShare = Number(amount / participants.length).toFixed(2);
    const newBalance = participants.map((participant) => {
      const balance = Number(fairShare - participant.paid).toFixed(2);
      return {
        name: participant.name,
        paid: participant.paid,
        balance: balance > 0 ? ` owes ${balance} $` : ` is owed ${-balance} $`,
      };
    });
    setResults(newBalance);
    const newBill = { title, amount, participants: newBalance };
    addBill(newBill);
  }
```
the details page:
```
export default function BillPageDetails() {
  const router = useRouter();
  const { id } = router.query;

  const { data: bill, error, isLoading } = useSWR(`/api/bills/${id}`);
  if (isLoading || error) return <h2>Loading Bill..</h2>;

  return (
    <BillDetails title={bill.title} total={bill.amount}>
      {bill.participants.map((participant, index) => (
        <li key={index}>
          {participant.name} paid {participant.paid} $, and{" "}
          {participant.balance}.
        </li>
      ))}
    </BillDetails>
  );
}
```
in my _app.js i have the states lifted and the function to add a new bill:
```
export default function App({ Component, pageProps }) {
  const [bill, setBill] = useState(null);
  const router = useRouter();

async function addBill(bill) {
    await fetch("api/bills", {
      method: "POST",
      body: JSON.stringify(bill),
    });
    setBill(bill);
    router.push("/bills/${id}");
  }
```
not sure if relevant, but my bill scheme looks like this:
```
const billSchema = new Schema({
  title: String,
  amount: { type: Number },
  id: String,
  participants: [
    {
      id: String,
      name: String,
      paid: { type: Number },
    },
  ],
  balance: { type: Number },
});
```
my [PR](https://github.com/d-opri/Kroko-Capstone/pull/11/files) 

thanks in advance for any help!!!

**Comment by mbosselmann** [comment, marked as answer]
Hi Danie,
the basic problem is that you want to work with an id, but when you create a new bill object you do not set an id. You can use `uuid` in your createBill function for this:

```js

async function createBill(bill) {
  await connectDatabase();

  const createdBill = await Bill.create({
    ...bill,
    id: uuidv4(),
  });

  return createdBill;
}
```

Another problem is your `addBill` function in your `CreateBill` page:
You send a POST request and afterwards you want to send the user directly to the details page of the newly created bill. If you want to do this you need to handle the response from to extract the id. Otherwise you do not know the new if of the createdBill:

```js
  async function addBill(bill) {
    const response = await fetch("api/bills", {
      method: "POST",
      body: JSON.stringify(bill),
    });

    const createdBill = await response.json();
    router.push(`/bills/${createdBill.id}`);
  }
```

Further remarks:
- The `addBill` function in your `_app.js` is not needed anymore and can be removed.
- You use react-uuid to create new ids. We would recommend to use a more established [package like `uuid`](https://www.npmjs.com/package/uuid) or you can also use `crypto` ([see this example by @shebtastic for crypto](https://github.com/shebtastic/demo-web-hh-22-7-neue-fische-fische-shop/blob/main/helpers/db.js)).

I hope this does help. :)

**Reply by d-opri** [reply]
Thank you mareike for all the help. i changed to uuidv4 but my details page still doesnt show the balance. i have balance as type number, but maybe i return a string? could that be? i tried this
```
     return {
        name: participant.name,
        paid: participant.paid,
        balance:
          balance > 0
            ? ` owes ${Number(balance)} $`
            : ` is owed ${-Number(balance)} $`,
      };
    });
```

but that didnt work, should i try in a different place/different method?

thank u for saving me
